docs/4.4/api-reference.md

```diff
@@ -113,50 +113,200 @@ func main() {
+
+*Rules*: A rule consists of two parts: the resources and verbs. Here's an example of a rule describing "read only" verbs applied to the SSH `sessions` resource. Depending on if it's under `Allow` or `Deny`, it means "allow/deny users of this role the ability to read or list active SSH sessions".
+
+```go
+services.NewRule(
+  services.KindSession, // check services.Kind* to see what other resources there are.
+  services.RO(), // helper function to get 'read only' verbs ("list" and "read")
+)
+```
+
+*NodeLabels*: Node labels are arbitrary key:value pairs that can be used to differentiate nodes by key attributes. For example, one of the most common label would be `environment`, which can be one of `development`, `staging`, or `production`.
+
+```go
+// Putting this label under "Allow" would allow the role to ssh into nodes with the "development" or "staging" labels.
+services.Labels{
+  "environment": utils.Strings{"development", "staging"},
+}
+```
+
+### Retrieve Role
+
+This is the equivalent of `tctl get role/admin`.
+
+```go
+role, err := client.GetRole("admin")
 if err != nil {
   return err
 }
 ```
 
-**Create a new role**
+### Create Role
+
+You can use the UpsertRole RPC to programmatically create a new role. This is the equivalent of `tctl create -f auditor-role.yaml`, where the `-f` flag signals to overwrite the auditor role if it exists already.
+
+For example, suppose you wanted to create a role for an auditor that could view all sessions, but could not access any servers.
+
+Using tctl, you would first create a role that allows reading and listing of the session resource like below. In addition, the user is explicitly denied access to all nodes in the deny block.
+
+```
+$ cat << EOF > /tmp/auditor-role.yaml
+kind: role
+version: v3
+metadata:
+  name: auditor
+spec:
+  options:
+    max_session_ttl: 8h
+  allow:
+    rules:
+    - resources: [session]
+      verbs: [list, read]
+  deny: {}
+    node_labels: '*': '*'
+EOF
+$ tctl create -f /tmp/auditor-role.yaml
+```
+
+To do something similar with the API:
 
 ```go
-// create a new auditor role which has very limited permissions
 role, err := services.NewRole("auditor", services.RoleSpecV3{
   Options: services.RoleOptions{
     MaxSessionTTL: services.Duration(time.Hour),
   },
   Allow: services.RoleConditions{
     Logins: []string{"auditor"},
     Rules: []services.Rule{
-      // - resources: ['session']
-      //   verbs: ['list', 'read']
       services.NewRule(services.KindSession, services.RO()),
     },
   },
   Deny: services.RoleConditions{
-    // node_labels: '*': '*'
     NodeLabels: services.Labels{"*": []string{"*"}},
   },
 })
 if err != nil {
   return err
 }
-
 if err = client.UpsertRole(ctx, role); err != nil {
   return err
 }
 ```
 
-**Update role**
+### Update Role
+
+The UpsertRole RPC can also be used to update an existing role.
+
+You can update practically any field within the role struct using the role's setter functions.
```

Comment:
It would be better to give a precise list of fields you can/can't update rather than saying "practically any".

Reply:
I think I'm going to change the wording rather than making an exhaustive list.